docs/roslyn_best_practices.md

```diff
@@ -0,0 +1,189 @@
+
+## Best Practices for Roslyn Analyzers and Code Fixes
+
+### 1. Use the *Microsoft.CodeAnalysis.Testing* framework
+
+Separation of analyzer and code fix tests increases complexity and code duplication and tends to decrease the overall confidence in the test suite. If you're testing entire files at a time, you will either be quickly overwhelmed by the number of files per test scenario or be tempted to put multiple test scenarios into a single file which shifts from unit to integration testing. The *Microsoft.CodeAnalysis.Testing* framework addresses these concerns.
+
+**Do**
+* Read the testing overview: [Microsoft.CodeAnalysis.Testing](https://github.com/dotnet/roslyn-sdk/blob/main/src/Microsoft.CodeAnalysis.Testing/README.md)
+
+### 2. Avoid member variables and state management
+Expect your analyzer to be invoked repeatedly and asynchronously. Design your analyzer so that execution can start processing a 2nd call before processing finishes for the 1st call. 
+
+**Do**
+* Pass information between methods via method arguments.
+
+**Do Not**
+* Do not use member variables to store instance data.
+
+### 3. Use abstractions and focus on the intent of your Analyzer rather than Roslyn
+Roslyn is a rich framework of information that describes every detail of code in every file of every project. The concepts can become overwhelming. Use abstractions to develop class names and methods that sharpen the focus on what the analyzer does by hiding how it achieves the goal.
+
+**Do**
+* Use extension methods, and wrapper objects to describe "what" the code does instead of "how" the code behaves.
+
+## Best Practices for Roslyn Analyzers
+
+### 1. "Bail out" Quickly
+Roslyn sees class files as rich trees of information. There will likely be millions of syntax nodes to evaluate across a large solution. You should limit the performance impact of running the analyzer by reducing the number of operations performed.
+
+**Do**
+* Use the [Syntax Visualizer](https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/syntax-visualizer) to build syntax. Where possible use specific syntax options. As an example, if you are building an analyzer that will examine class inheritance then you should use `SyntaxKind.BaseList`, which will occur much less often than the `SyntaxKind.IdentifierName`.
+* Use information already available from `SyntaxNodeAnalysisContext` to quickly filter relevant information. As an example, if you build an analyzer that looks for a method then you would be evaluating `MemberAccessExpressionSyntax` nodes. In this scenario, you should also check the parent of the node to figure out if this Syntax is a method or a property.
+
+**Do Not**
+* Do not put string compare operations above other conditional checks that can be performed. String comparison is slower than checking for a type or examining an enumeration value.
```

Comment:
I think the important thing is not so much string comparison, but calling `.ToString()` and comparing that output. Comparing the ValueText of an identifier is probably cheaper than ISymbol.ToString

Reply:
Oh this may be blurring the next issue it seems